security/pkg/k8s/configmap/configmap.go

```diff
@@ -74,6 +80,31 @@ func (c *Controller) InsertCATLSRootCert(value string) error {
 	return nil
 }
 
+// InsertCATLSRootCertWithRetry updates the CA TLS root certificate in the configmap with
+// retries until timeout.
+func (c *Controller) InsertCATLSRootCertWithRetry(value string, retryInterval,
```

Comment:
@costinm @rlenglet @myidpt should we worry about atomicity here?  Is it possible for this operation to partially fail and corrupt the current Root CA in the process?

Reply:
Good catch.
This partial failure will cause the CSR connections from Node Agent fail after the old root cert expires (although Istio addon doesn't support SDS). Current routine to call this function is problematic because it's called only when the root cert rotation happens - if this function fails due to temp error, there's no more chance to sync the root cert in ConfigMap with the cert in the root secret, unless Citadel is restarted.  
One way to fix it is to periodically sync the root cert in the ConfigMap with the root cert in the root keycert secret, in the routine checkAndRotateRootCertForSigningCertCitadel(), as I commented in the following.

Reply:
You are right, there's a problem. `checkAndRotateRootCertForSigningCertCitadel` first updates the private key in the secret, then updates the cert in the config map. If the config map update fails, the private key / secret is not reverted. This is a problem because the same `checkAndRotateRootCertForSigningCertCitadel` only checks the secret, not the cert, so it can't detect the inconsistency and fix it.

One way to fix it is to have `checkAndRotateRootCertForSigningCertCitadel` load both the cert and the key, and if they are inconsistent or they are about to expire, then update them. Shouldn't be too hard to fix?!

Reply:
We came to similar conclusions. :)

Reply:
About the atomicity concern of the "update" operation itself (in case @duderino was actually asking about this):
I was looking at https://kubernetes.io/docs/reference/using-api/api-concepts/
I didn't find explicit "atomicity" in it about the API operations, but I think K8s master will roll back the operation if the update process is corrupted. Otherwise there will be big issues on K8s.

Reply:
The problem here is that there are 2 resources to update: the key in the secret, and the cert in the config map. There is no way in the k8s API to update both resources atomically. And this PR doesn't properly handle the case when the cert update fails.

Reply:
So I left the comment to just periodically sync the secret with the ConfigMap (by default every hour). This should work, and the thought is:

1. We should rotate the secret if it's expiring, no matter whether ConfigMap failed to be updated, since that's needed for both file mount and SDS.
2. If ConfigMap failed to be updated, the auto-rotator tries to update it periodically (by default every hour). As long as it's successfully updated before the old root cert expires, it doesn't break the SDS flow.

Reply:
No it's not ok. The node agent gets the root CA cert from that config map, right?
So it will create and send CSRs against that root CA cert. And I don't think that will work.

Reply:
You should try reverting to a consistent state if you can't update both the secret and config map in a run.

Reply:
Sorry I think there's a misunderstanding. Let me clarify it.

The node agent only uses the root cert in the config map to authenticate the Citadel's server cert. Even though the cert in the config map is the old one, the Citadel will propagate the new root cert to node agent through the CSR response, and the node agent will forward the new root cert to Envoy as it's root of trust.

So if the root cert in the config map is the old one, in the worst case, when the root cert expires the node agent will not able to talk to Citadel. It's unlikely to happen if Citadel syncs the config map with the secret every hour (as I proposed).

Reply:
That's the conclusion I wanted to get to. For one hour, no pod could start, because node agent couldn't connect to Citadel because it can't verify its certificate. A 1-hour outage is not OK.

Reply:
I think unless the root cert expires (which is unlikely within 1 hour), the old cert can still be used to verify the new server TLS cert - because they are using the same root private key.
We can talk tomorrow as well :)

Reply:
I have verified that the Citadel agent won't be able to connect Citadel with the old cert even if it's not expired (unexpected, good we verified it :( ). So Romain is right. Let's roll back the root ca secret if we fail to update the ConfigMap.

Reply:
@duderino thanks for pointing this out. I have confirmed that the update call is an atomic operation. https://stackoverflow.com/a/58546663/8922031

Reply:
I have added the rollback step and unit test for the rollback.